### ISSUE TITLE
Optionality of restriction base

### ISSUE THREAD

**Issue by CBenghi** [buildingSMART/IDS#195]
As noted by @pasi-paasiala,
we need to agree whether base type restrictions are required to pass the audit.

Is the following restriction acceptable?

```
<ids:entity>
	<ids:name>
		<xs:restriction>
			<xs:enumeration value="IFCSTAIR"/>
			<xs:enumeration value="IFCSTAIRFLIGHT"/>
		</xs:restriction>
	</ids:name>
</ids:entity>
```

If not, a test case needs to be created and the audit tool should be updated accordingly.

**Comment by MarcelStepien** [comment]
In my opinion the base-restriction is obsolete, since it is indirectly inferred by the datatype/measure in a property-facet (**IFCTEXT** checks for **xs:string**, **IFCREAL** checks for **xs:double**). In all other facets, such as in your example of checking entity-name, setting the base restriction to anything other then **xs:string** makes no sense. 

One could argue that alphanumerical nomenclature can be excluded by simply restricting the base to something like **xs:integer**. However, this requirement can and should be solved using **RegEx pattern** instead.

**Comment by aothms** [comment]
I agree, this XSD base thing only increases confusion, because of the additional indirection in the mapping between IFC and XSD types. And what to do in the corner cases: e.g ifcdatetime is a str in ISO 8601 format is it XSD string or XSD date? Disclaimer: I haven't been too involved in recent discussions, I could be confusing things.

**Comment by pasi-paasiala** [comment]
I agree that there shouldn't be two ways to do the same thing. Here the question was about IFC entity. Should we add a new data type IfcClass (or something)? Also we could make the data type non-optional.

**Comment by mysterieux30** [comment]
That is true, this base attribute is bringing me into a real confusion like in this example:
```
         <ids:name>
            <ids:simpleValue>TreadLength</ids:simpleValue>
          </ids:name>
          <ids:value>
            <xs:restriction base="xs:string">
              <xs:pattern value="^(0*(\.\d+)|[1-9]\d*(\.\d+)?)$" />
            </xs:restriction>
          </ids:value>
```

In this case, for example, the value must have a predefined type that is based on simple type "string":
like:
**IfcIdentifier**: Defined type of simple type _STRING_ for identification purposes.
**IfcLabel**: Defined type of simple type _STRING_ for naming purposes.
**IfcText**: Defined type of simple type _STRING_ for descriptive purposes.
**IfcDateTime**: Defined type of simple type _STRING_ to represent a date and time.
**IfcDate**: Defined type of simple type _STRING_ to represent a date.
**IfcTime**: Defined type of simple type _STRING_ to represent a time.
**IfcDuration**: Defined type of simple type _STRING_ to represent a duration.



In case the value have a predefined type that is based on simple type other than "string", the validation should return an error?
example:
**IfcInteger**: Defined type of simple type _INTEGER_.
**IfcReal**: Defined type of simple type _REAL_.
**IfcBoolean**: Defined type of simple type _BOOLEAN_.
**IfcLogical**: Defined type of simple type _LOGICAL_.
**IfcTimeStamp**: Defined type of simple type _INTEGER_.

**Comment by mysterieux30** [comment]
In case the restriction is added to an **attribute facet** / value, the type of the values are already defined in the Ifc schema, the use of base may sometimes bring the user into getting always an error when for example the predefined type is "IfcLabel" and the base has a value like "xs:integer".
Sorry for the disturbance, I'm just trying to share my thoughts with you.

**Comment by Moult** [comment]
@mysterieux30 the restriction base has no impact on the IFC data type. For attributes the IFC data type is fixed. For properties, the data type is specified using the `datatype` attribute. The restriction base of string in your example simply refers to the pattern value being string matching.

**Comment by mysterieux30** [comment]
[CBenghi](https://github.com/CBenghi) no i checked the XSD Schema using the software Altova, the base attribut is marked as optional

**Comment by CBenghi** [comment]
@mysterieux30 you are right, but we checked the XML schema, and we think this is a problem with ALTOVA.

**Comment by CBenghi** [comment]
The base attribute on restriction must be specified to comply with XML schema.
I've added checking to the audit tool (upcoming release).

**Comment by BIMvoice** [comment]
I may be misreading this, but I could not reproduce the schema behaviour the conclusion rests on, so I wanted to put the evidence up rather than sit on it.

`idsValue` does not derive a type from `xs:restriction`. It references the element:

```xml
<xs:element ref="xs:restriction" minOccurs="1" maxOccurs="1"/>
```

That binds to the standalone global `restriction` element in the schema-for-schemas, where `base` is declared `use="optional"`. The `base="required"` version is on `restrictionType`, which is the complex type used when `xs:restriction` appears inside an actual type derivation, and that is not how `ids.xsd` uses it.

Checked two ways independently, with `xmlschema` (Python) and `libxml2` via `xmllint`: an `.ids` whose `xs:restriction` has no `base` attribute validates against `Schema/ids.xsd` on both. To rule out the validator simply skipping that subtree, I ran a control with a made-up facet name inside the restriction, and both engines correctly reject it. So the content model is genuinely being checked, and `base` is genuinely optional there.

That would also mean @mysterieux30's Altova observation earlier in this thread was correct rather than a tool quirk.

Two things I want to be careful about. The element's own annotation says *"base attribute and simpleType child are mutually exclusive, but one or other is required"*, which is a real argument for wanting `base` mandatory, it just is not something plain XSD 1.0 enforces structurally, and IDS never uses the `simpleType` child form. And the audit tool's check appears to be partial: the relevant test is marked `[Obsolete]` and the matcher carries its own `// todo: discuss with group: do we want to force the base type requirement for strings?`.

So this may be a case where documenting the rule is not enough, because the schema as shipped does not carry it. Two ways forward, both your call as schema author:

1. Document current behaviour and strongly recommend always specifying `base`. I have drafted that, happy to open it.
2. Tighten `ids.xsd` so `base` is structurally required, for example by giving `idsValue`'s restriction its own locally defined complex type rather than a bare `ref`, which would make the audit tool's intent actually enforced.

Happy to be told I have misread the schema, and glad to adjust the doc either way once you have weighed in.
